### PR TITLE
Startup action

### PR DESCRIPTION
**Pull request by the-Bruce**
Creates a startup action that can pass data into the main commands in a Reader.
This also incorporates the previous work on plugin disabling and reloading (I'd expected that to be merged by now, but schedules conspired against). 
Note, this changes the Monad stack, but introduces some utility features to help mitigate that in the future.
It also quite significantly changes the back-end, but hopefully this is mostly transparent to Plugin authors.

You can see an example of it in use with the welcome plugin

### CODE REVIEW

File: src/Tablebot/Plugins/Administration.hs [hunk 1]
```diff
@@ -0,0 +1,180 @@
...
+blacklistComm (WErr (Right (_))) = listBlacklist
+
+addBlacklist :: String -> Message -> DatabaseDiscord SS ()
+addBlacklist pLabel m = requirePermission Superuser m $ do
+  known <- ask
+  -- It's not an error to add an unknown plugin (so that you can pre-disable a plugin you know you're about to add),
+  -- but emmit a warning so people know if it wasn't deliberate
+  when ((pack pLabel) `notElem` known) $ sendMessage m "Warning, unknown plugin"
+  extant <- exists [PluginBlacklistLabel ==. pLabel]
+  if not $ extant
```
Comment (finnbar) [inline src/Tablebot/Plugins/Administration.hs:60]: I don't think this `$` is needed?

File: src/Tablebot/Plugins/Administration.hs [hunk 2]
```diff
@@ -0,0 +1,180 @@
...
+  sendMessage m (format pl (blacklisted bl))
+  where
+    allBlacklisted :: [Filter PluginBlacklist]
+    allBlacklisted = []
+    -- Oh gods, so much formatting.
+    format :: [Text] -> [Text] -> Text
+    format p bl = "**Plugins:**\n```\n" <> (T.concat $ map (format' len' bl) (filter (disableable . T.uncons) p)) <> "```" <> unknown (filter (`notElem` p) bl)
+      where
+        len' :: Int
+        len' = maximum $ map T.length p
```
Comment (finnbar) [inline src/Tablebot/Plugins/Administration.hs:90]: This isn't very clear to me - I get that it's providing a formatted version, but it isn't very clear how. (The names don't help a lot here - e.g. what does `format'` do?) It should at least be multilined (newline after some `<>`).

**Review by finnbar — CHANGES_REQUESTED** [review]
Right, this is a big commit so forgive me for the many suggestions. I'm properly impressed by this code, to the point where I was thinking "holy shit, Anna is a far better Haskell programmer than I am", but I still have a few critiques. Please note that this is currently stylistic as I'm not going to have time to test this code for a little bit, so my points may have really simple answers that I've just missed.

One overarching design decision that I suggested in Discord and I'll restate here: instead of having user computations be of type `DatabaseDiscord env res`, instead name it `DatabaseDiscordEnv env res` and then have `type DatabaseDiscord res = DatabaseDiscordEnv () res`. This means you don't have to change any of the plugins (except for adding the Tablebot Database file import) in most cases. You may also want `CommandWithEnv env` in the same vein, so that `type Command = CommandWithEnv ()` and again no changes are needed.

Otherwise, this is great!

File: src/Tablebot/Handler/Administration.hs [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+
+share
+  [mkPersist sqlSettings, mkMigrate "adminMigration"]
+  [persistLowerCase|
+PluginBlacklist
+    label String
+    deriving Show
+|]
+
+currentBlacklist :: SqlPersistM ([Text])
```
[the remark this answers is not in the document]
Comment (the-Bruce) [reply src/Tablebot/Handler/Administration.hs:21]: Gone

File: src/Tablebot/Handler/Administration.hs [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+currentBlacklist = do
+  bl <- selectList allBlacklisted []
+  return $ fmap (pack . pluginBlacklistLabel . unentity) bl
+  where
+    allBlacklisted :: [Filter PluginBlacklist]
+    allBlacklisted = []
+    unentity :: (Entity a) -> a
+    unentity a = entityVal a
```
[the remark this answers is not in the document]
Comment (the-Bruce) [reply src/Tablebot/Handler/Administration.hs:29]: Oops. That was a result of fighting the type system (I often breakout just the type signature I'm aiming for, and in this case forgot to remerge...)

File: src/Tablebot/Plugins/Administration.hs [hunk 5]
```diff
@@ -0,0 +1,180 @@
+-- |
+-- Module      : Tablebot.Plugins.Administration
+-- Description : A command that outputs its input.
```
[the remark this answers is not in the document]
Comment (the-Bruce) [reply src/Tablebot/Plugins/Administration.hs:3]: Updated (`b9b5bd9`)

File: src/Tablebot/Plugins/Administration.hs [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+
+-- | @SS@ denotes the type returned by the command setup. Here its unused.
+type SS = [Text]
+
+blacklist :: Command SS
+blacklist = Command "blacklist" (parseComm blacklistComm)
+
+blacklistComm ::
+  WithError
+    "Unknown quote functionality."
```
[the remark this answers is not in the document]
Comment (the-Bruce) [reply src/Tablebot/Plugins/Administration.hs:39]: Done (`a0790de`)

File: app/Main.hs [hunk 7]
```diff
@@ -8,7 +9,7 @@ import Tablebot (runTablebot)
 import Tablebot.Plugins (plugins)
 
 main :: IO ()
-main = do
+main = forever $ do
```
[the remark this answers is not in the document]
Comment (the-Bruce) [reply app/Main.hs:12]: Done (`e39c95f`)

File: src/Tablebot/Plugins/Administration.hs (hunk 2 above)
Comment (the-Bruce) [reply src/Tablebot/Plugins/Administration.hs]: Take a look at it now. I've hopefully cleaned it up a fair bit.

File: src/Tablebot/Plugins/Administration.hs (hunk 1 above)
Comment (the-Bruce) [reply src/Tablebot/Plugins/Administration.hs]: Fixed (d7540a8)

**Review by finnbar — APPROVED** [review]
Tested and all is good.

**Pull request merged**